designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+this interface will need to change accordingly. However, the CmdStan
+grammar is intended to be quite stable. Moreover, it is not
+necessarily the case that radical changes to the CmdStan grammar could
+be hidden behind something other abstraction.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Other than the direct representation of the CmdStan syntax tree, the
+proposal contains nothing new. Utilizing a concrete representation of
+the syntax tree does have benefits:
+- all outputs handled via single type `CmdStanOutput`
+- elimination of individual structures and methods for each of Stan's
+inference algorithms
+
+The question remains: is this is a good idea?
+
+## Argument tree considerations
+
+As described above, the grammar for CmdStan arguments passed at the
+command line can be represented as a syntax tree through the use of
+sum and product types.
+- This enables compile-time validation of argument consistency -- the
+  worst that can happen is you provide a value that CmdStan does not
+  like (e.g. `num_threads=-20`).
+- At minimum, this will move a variety of run-time errors to compile
+  time; it might even help users to understand the methods and options
+  CmdStan provides.
+- This enables re-use of a parameterized argument tree -- one could
+  replace the inference method while leaving the other options
+  (e.g. data files) constant. As shown in [this
+  example](https://github.com/andrewjradcliffe/cmdstan-rs/blob/main/examples/bernoulli-many/main.rs),
+  such an approach can be quite expressive.
+
+Furthermore, representation as a concrete syntax tree enables the
+possibility of interesting features. One could parse the syntax tree
+from:
+- a string written to a log file
+- a string which is consistent with the grammar that CmdStan accepts
+
+The latter is interesting in that a user's extant command line input
+is all that is required to use the Rust interface. For example,
+this leads to the following syntax:
+
+```rust
+// Assuming we implemented this through the `FromStr` trait
+let tree: ArgumentTree = "method=sample data file=bernoulli.data.json".parse().unwrap();
+```
+
+This would substantially lower the barrier to adoption of the Rust
+interface as the user need only know what they are already doing.
+
+Due to Rust's orphan rules, such features would need to be implemented
+within this crate; they could be placed behind a feature gate to
+minimize compile time. It stands to reason that if we can translate to
+a string, we should be able to perform the inverse operation.
+
+The design philosophy here would be: a valid parse is whatever CmdStan
+is willing to accept. However, CmdStan accepts some weird statements.
+For example:
+```bash
+./bernoulli method=sample adapt engaged engaged=0 engaged engaged=1 gamma engaged gamma \
+    data file=bernoulli.data.json
+```
+
+The proposal is to use [pest](https://github.com/pest-parser/pest),
```

Comment:
One thing I am curious about is total dependencies (it may be worth calling this out as its own list/table somewhere). 

Both Python and R's wrappers at least strive to be minimal, within what their standard libraries provide. I think the wonder of `cargo` making dependencies on the whole less painful has a downside of them tending to proliferate. 

Could items like this be hidden behind a cargo feature? Continuing the only-pay-for-what-you-want mentality from earlier

Reply:
Things like MSRV may also be worth mentioning!

Reply:
I don't have any objections to `thiserror` - it seems quite common in practice and rather small in terms of scope.

You bring up an interesting other point: I (sort of implicitly) expected that this wrapper could be written using entirely safe Rust. Is that `unsafe` function really needed, or is it called as an optimization?

What other usages of `unsafe` would you anticipate, and why?

Reply:
The single unsafe function arose from the need to split the string on the occurrence of `'.'`, with the objective of replicating CmdStan's interesting treatment of paths. This could easily be eliminated if we weren't dealing with Os-specific details (hidden behind [WTF-8](https://en.wikipedia.org/wiki/UTF-8#WTF-8) in OsString).

CmdStan basically just splits at the first occurrence of `'.'`, starting from the right. 
This becomes a problem when the only thing to the left of the first `'.'` is the empty string and there the input matches any of the forms below.

```bash
./bernoulli id=2 method=sample num_chains=3 data file=bernoulli.data.json output file='..'
# produces as output file names: `._2.`, `._3.`, `._4.`

./bernoulli id=2 method=sample num_chains=3 data file=bernoulli.data.json output file=.foo
# produces as output file names: `_2.foo`, `_3.foo`, `_4.foo`

# this is actually a repeat of the above, with "foo" as the empty string
./bernoulli id=2 method=sample num_chains=3 data file=bernoulli.data.json output file='.'
# produces as output file names: `_2.`, `_3.`, `_4.`
```

It is not possible to easily treat these by two edge cases due to the [definition of an extension](https://doc.rust-lang.org/std/path/struct.Path.html#method.extension).
I might be able to come up with another way to do it, but it will complicate the code accordingly. 

I see no issue with `unsafe` which is limited to a very specific use case that is fundamentally sound. This use is particularly easy to verify soundness: the `&[u8]` slices are created by [`as_encoded_bytes`](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.as_encoded_bytes) and promptly consumed; when the byte slices are re-joined, it is impossible for either to have become invalid UTF-8.

It is worth noting that `OsStr::from_encoded_bytes_unchecked` only appears in two places; both appearances involve inputs which are created within the same function body, that is, no foreign inputs, no intermediates escape the scope in which they are created and said intermediates are consumed in the same function body. Incidentally, it is more efficient to play with byte slices, but that was not my objective.

Each use looks like:
```rust
let bytes = path.as_encoded_bytes();
let mut iter = bytes.rsplitn(2, |b| *b == b'.');
let (prefix, suffix) = match (iter.next(), iter.next()) {
    (Some(suffix), Some(prefix)) => {
        // SAFETY:
        // - each fragment only contains content that originated
        //   from `OsStr::as_encoded_bytes`.
        // - split with ASCII period, which is a non-empty UTF-8
        //   substring
        unsafe {
            (
                OsStr::from_encoded_bytes_unchecked(prefix),
                OsStr::from_encoded_bytes_unchecked(suffix),
            )
        }
    }
    _ => (path, "csv".as_ref()),
};
```

I general, my philosophy on use of `unsafe` is that it is best to avoid where possible; for a crate like this, I certainly wouldn't expose any `unsafe` public APIs. There's no reason to expose an unsafe internal API for this crate, as there is nothing that warrants it here.

I realize that [my response above](https://github.com/stan-dev/design-docs/pull/51#discussion_r1413267023) could actually be read as somewhat alarming -- potentially implies lots of `unsafe`. Limited to two locations ([1](https://github.com/andrewjradcliffe/cmdstan-rs/blob/42a4e0be0b6669096a81bcbc514878beb4e57cb5/src/argument_tree.rs#L76), [2](https://github.com/andrewjradcliffe/cmdstan-rs/blob/42a4e0be0b6669096a81bcbc514878beb4e57cb5/src/argument_tree.rs#L177C13-L177C13)) with no plans to expand beyond these, I hope that the above is reassuring.

Reply:
This also sounds like something we will probably change (https://github.com/stan-dev/cmdstan/issues/1213) - the current system is mostly designed-by-accident

Reply:
1. Your use of `unsafe` seem fine to me. However, since essentially the same unsafe code is used in two places I'd encapsulate it in a single safe-to-call helper function.
2. This file extension handling corner case feels like a CmdStan bug and I wouldn't bother supporting it beyond a clear error message.
3. It is quite frustrating that `as_/from_encoded_bytes` does not provide a _safe_ API for splitting `OsStr`s. Nevertheless, in this case `unsafe` can be avoided with a couple of allocations:
```rust
use std::ffi::{OsString, OsStr};
let (prefix, suffix) = match (path.file_stem(), path.extension()) {
    (Some(prefix), Some(suffix)) => (prefix, suffix.to_os_string()),
    (Some(name), None) if name.as_encoded_bytes().first() == Some(&b'.') => {
        let mut fake_name = std::path::PathBuf::from("x");
        fake_name.as_mut_os_string().push(name);
        (OsStr::new(""), fake_name.extension().unwrap().to_os_string())
    }
    (Some(name), None) =>
        (name, OsString::from("csv")),
    (None, _) => {
        // No file name, which means the last path component is either "." or ".."
        if path.as_os_str().as_encoded_bytes().ends_with(b"..") {
            (OsStr::new("."), OsString::from(""))
        } else {
            (OsStr::new(""), OsString::from(""))
        }
    }
};
```
4. Coming back to point (2), if the user specifies `output file=out.d/file` CmdStan tries to write to `out_1.d/file` _etc_ but if the directory `out_1.d` does not exist CmdnStan won't create it and the output just silently disappears. That can't be right.

Reply:
In order:
1. I completely agree. It should be moved to a single function.
2. I agree again. I found this when I was writing tests for the Rust code. However, if CmdStan does it, it seems fine to support it if it is easy enough. Splitting WTF-8 encoded bytes on ASCII is fundamentally sound, so the `unsafe`-ness doesn't bother me. In fact, `std` uses the [same approach](https://doc.rust-lang.org/src/std/path.rs.html#318); I had not looked at `std` until now.
3. It is only the conversion back to `OsStr` that is `unsafe`, but I agree that a safe API would be nice. Unfortunately, literal split on the first period from the right is very hard to avoid. Your approach is clever, but is defeated by inputs such as below:
```bash
function bernoulli() {
    ./bernoulli data file=bernoulli.data.json sample num_chains=3 output file=$1
}

mkdir -p foo/bar
                            # files created by CmdStan
bernoulli 'foo/.bar'        # "foo/_1.bar", "foo/_2.bar", "foo/_3.bar"
bernoulli 'foo/bar/'        # "foo/bar/_1.csv", "foo/bar/_2.csv", "foo/bar/_3.csv"
bernoulli 'foo/bar/.'       # "foo/bar/_1.", "foo/bar/_2.", "foo/bar/_3."
bernoulli 'foo/bar/..'      # "foo/bar/._2.", "foo/bar/._3.", "foo/bar/._4."
bernoulli 'foo/bar/...'     # "foo/bar/.._2.", "foo/bar/.._3.", "foo/bar/.._4."
```
One might think that `path::parent` and `path::file_name` would reduce the problem to the code you gave, substituting the result of `path::file_name`, but that does not work due to path normalization. For example, `foo/bar/.` is equivalent to `foo/bar` when treated as a path; likewise for `foo/bar/`.

From a code complexity standpoint, splitting at the `'.'` byte is much simpler, as it does not require addition of more case logic, and matches what CmdStan does by default. 

This is a good example of where `unsafe` is both easier to maintain and verify the correctness of. Using the `std::path` API quickly becomes cumbersome, as it performs the normalization of path components one would naturally expect; in essence, the logic of the program must now test for all the ways that a string with strange placements of `'.'` would be subject to path normalization, then avoid them.

4. This was sufficiently interesting that I wrote it as an [issue](https://github.com/stan-dev/cmdstan/issues/1226) for the CmdStan repository.

----
### Summary

Until CmdStan switches from splitting on the last dot, there is no reason to deviate from said behavior in the Rust treatment of paths. Path normalization is not trivial (see [cppreference](https://en.cppreference.com/w/cpp/filesystem/path) for a nice recapitulation of the steps). Effort spent on attempting to reproduce the CmdStan behavior via a convoluted use of `Path` and `OsStr`, on the basis that no code should be `unsafe`, is likely to lead to logical errors-- I am not keen on working out all pathological (no pun intended) behavior induced by lack of path normalization and splitting on the last `'.'`. It is much simpler and maintainable to verify that the invariants of an `unsafe` function are satisfied than to create a [Rube Goldberg machine](https://en.wikipedia.org/wiki/Rube_Goldberg_machine).

Examples like [Case 2](https://github.com/stan-dev/cmdstan/issues/1226) demonstrate that such a machine would not be feasible, as it is not sufficient to split parent and child, check the respective start/end, and be done. One must scan the entire string for dots and this is only possible after `as_bytes_encoded`, hence, we are back to `unsafe` to put the halves back together (i.e. right back to the original proposal).

Furthermore, the discrepancy between CmdStan's behavior and the proposed interface would be a a potential source of confusion. If someone wants to use wacky file names and the program (CmdStan) accepts them, we would be remiss to impose artificial limitations.